File: app/test/dispatch_test.dart
// How much of an order line remains to be dispatched.
//
// Checked against `shared/fixtures/dispatch.json`, the same file
// `client/src/domain/__tests__/dispatch.test.ts` reads.

import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

import 'package:manna_field_sales/core/dispatch.dart';

Map<String, dynamic> _fixture() {
  // Relative to `app/`, which is where `flutter test` runs from.
  final f = File('../shared/fixtures/dispatch.json');
  return json.decode(f.readAsStringSync()) as Map<String, dynamic>;
}

void main() {
  final fx = _fixture();

  group('shared fixture: remaining to dispatch', () {
    for (final raw in fx['remaining'] as List) {
      final c = (raw as Map).cast<String, dynamic>();
      test(c['why'] as String, () {
        final line = (c['line'] as Map).cast<String, dynamic>();
        final expect_ = (c['expect'] as Map).cast<String, dynamic>();
        final left = remainingToDispatch(line);
        expect(left.rolls, expect_['rolls']);
        expect(left.looseBelts, expect_['loose_belts']);
      });
    }
  });

  group('shared fixture: fully dispatched', () {
    for (final raw in fx['fully_dispatched'] as List) {
      final c = (raw as Map).cast<String, dynamic>();
      test(c['why'] as String, () {
        final line = (c['line'] as Map).cast<String, dynamic>();
        expect(isFullyDispatched(line), c['expect']);
      });
    }
  });
}

File: client/src/domain/__tests__/dispatch.test.ts
/**
 * How much of an order line remains to be dispatched.
 *
 * Checked against `shared/fixtures/dispatch.json`, the same file
 * `app/test/dispatch_test.dart` reads.
 */

import { describe, expect, it } from 'vitest';
import cases from '../../../../shared/fixtures/dispatch.json';
import { isFullyDispatched, remainingToDispatch } from '../dispatch';

describe('shared fixture: remaining to dispatch', () => {
  for (const c of cases.remaining) {
    it(c.why, () => {
      const left = remainingToDispatch(c.line);
      expect(left).toEqual({ rolls: c.expect.rolls, looseBelts: c.expect.loose_belts });
    });
  }
});

describe('shared fixture: fully dispatched', () => {
  for (const c of cases.fully_dispatched) {
    it(c.why, () => {
      expect(isFullyDispatched(c.line)).toBe(c.expect);
    });
  }
});
